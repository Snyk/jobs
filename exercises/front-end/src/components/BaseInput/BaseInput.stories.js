import '~/assets/scss/styles.scss';
import BaseInput from './BaseInput';

export default {
  title: 'Library/Base Components/Input',
  component: BaseInput,
};

export const Default = () => ({
  components: { BaseInput },
  props: {
    label: {
      default: 'Label text',
    },
    placeholder: {
      default: 'Write your text here',
    },
    state: {
      default: null,
    },
    size: {
      default: null,
    },
    inline: {
      default: false,
    },
    error: {
      default: false,
    },
    note: {
      default: '',
    },
  },
  template: `
    <BaseInput
      :label="label"
      :placeholder="placeholder"
      :state="state"
      :size="size"
      :inline="inline"
      :error="error"
      :note="note"
    />
  `,
});

export const States = () => ({
  components: { BaseInput },
  template: `
  <div>
    <BaseInput
      label=":hover"
      state="hover"
      placeholder="Write your text"
      id="input-2"
    />
    <BaseInput
      label=":focus"
      state="focus"
      placeholder="Write your text"
      id="input-3"
    />
    <BaseInput
      label="[disabled]"
      disabled
      value="Write your text"
      id="input-4"
    />
    <BaseInput
      label="Error"
      error
      placeholder="This is a required field"
      id="input-5"
    />
  </div>
`,
});

export const Notes = () => ({
  components: { BaseInput },
  template: `
  <div>
    <BaseInput
      label="With note"
      type="password"
      note="Passwords must contain at least one capital letter, a number, and a special character."
    />
    <BaseInput
      label="Error with note"
      error
      note="This is a required field"
    />
  </div>
`,
});

export const Types = () => ({
  components: { BaseInput },
  template: `
  <div>
    <BaseInput
      label="Text input"
      placeholder="Write your text"
      id="input-1"
    />
    <BaseInput
      type="number"
      label="Number input"
      placeholder="1-10"
      :value="9"
      id="input-9"
      :min="0"
      :max="10"
    />
    <BaseInput
      type="date"
      label="Date input"
      id="input-10"
    />
  </div>
`,
});

export const Inline = () => ({
  components: { BaseInput },
  template: `
  <BaseInput
    label="Inline"
    placeholder="your_org"
    id="inline-input-1"
    inline
  />
`,
});

export const Size = () => ({
  components: { BaseInput },
  template: `
  <div>
    <BaseInput
      label="Default size"
      placeholder="Write your text"
      id="input-1"
    />
    <BaseInput
      label="Small size"
      placeholder="Write your text"
      id="input-2"
      size="small"
    />
  </div>
`,
});
