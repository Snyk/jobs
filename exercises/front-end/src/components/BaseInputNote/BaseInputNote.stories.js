import '~/assets/scss/styles.scss';
import BaseInputNote from './BaseInputNote';

export default {
  title: 'Library/Base Components/InputNote',
  component: BaseInputNote,
};

export const Default = () => ({
  components: { BaseInputNote },
  props: {
    variant: {
      default: 'info',
    },
  },
  template: `
    <BaseInputNote :variant="variant">
      Base Input Note ({{variant}}).
    </BaseInputNote>
  `,
});

export const Info = () => ({
  components: { BaseInputNote },
  template: `<BaseInputNote variant="info">A info input note for informational purposes.</BaseInputNote>`,
});

export const Warning = () => ({
  components: { BaseInputNote },
  template: `<BaseInputNote variant="warning">A warning input note for when there's a non-critical problem.</BaseInputNote>`,
});

export const Success = () => ({
  components: { BaseInputNote },
  template: `<BaseInputNote variant="success">A success input note for when something awesome has happened.</BaseInputNote>`,
});

export const Danger = () => ({
  components: { BaseInputNote },
  template: `<BaseInputNote variant="danger">A danger input note for when something unpleasant has happened.</BaseInputNote>`,
});
