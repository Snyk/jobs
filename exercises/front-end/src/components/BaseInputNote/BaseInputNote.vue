<template>
  <div class="vue--input-note" :class="computedClasses">
    <slot />
  </div>
</template>

<script>
import { isInList } from '~/lib/prop-validators';

/**
 * A note to appear under an input
 */
export default {
  name: 'BaseInputNote',

  status: 'under-review',

  props: {
    /**
     * Variant of BaseInputNote. Must be one of `default`, `info`, `warning`,
     * `error`, `danger`, `success`. Defaults to `default`
     */
    variant: {
      type: String,
      default: 'info',
      validator: isInList([
        'default',
        'info',
        'warning',
        'error',
        'danger',
        'success',
      ]),
    },
  },

  computed: {
    computedClasses() {
      return {
        [`vue--input-note--${this.variant}`]: true,
      };
    },
  },
};
</script>

<style lang="scss" scoped>
@import '~/assets/scss/utils';

.vue--input-note {
  @include typeset(caption);

  &--default,
  &--info {
    color: color(base, dimmed);
  }

  &--warning {
    color: color(alert, warning-text);
  }

  &--error,
  &--danger {
    color: color(alert, danger-text);
  }

  &--success {
    color: color(alert, success-text);
  }
}
</style>
