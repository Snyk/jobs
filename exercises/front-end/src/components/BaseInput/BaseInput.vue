<template>
  <LayoutFormElement
    class="vue--input"
    :class="computedClasses"
    :inline="inline"
    :fill-width="true"
  >
    <template v-if="label" #label>
      <label class="vue--input__label" :for="$attrs.id" v-text="label" />
    </template>
    <input
      ref="input"
      :type="type"
      v-bind="$attrs"
      class="vue--input__field"
      :class="computedInputClasses"
      v-on="listeners"
    />
    <BaseInputNote v-if="note" class="vue--input__note" :variant="noteVariant">
      {{ note }}
    </BaseInputNote>
  </LayoutFormElement>
</template>

<script>
import LayoutFormElement from '~/components/LayoutFormElement/LayoutFormElement';
import BaseInputNote from '~/components/BaseInputNote/BaseInputNote';
import { isInList } from '~/lib/prop-validators';

/**
 * A super flexible input field with optional label.
 */
export default {
  name: 'BaseInput',

  status: 'under-review',

  components: { LayoutFormElement, BaseInputNote },

  inheritAttrs: false,

  model: {
    event: 'change',
  },

  props: {
    /**
     * Sets the default type for the input to `text`.
     */
    type: {
      type: String,
      default: 'text',
    },
    /**
     * The label of the form input field.
     */
    label: {
      type: String,
      default: null,
    },
    /**
     * Whether to show an error for this field
     * `true, false`
     */
    error: {
      type: Boolean,
      default: false,
    },
    /**
     * Manually trigger various states of the input.
     * `hover, active, focus`
     */
    state: {
      type: String,
      default: null,
      validator: isInList(['hover', 'active', 'focus']),
    },
    /**
     * Whether the form input field is displayed inline.
     * `true, false`
     */
    inline: {
      type: Boolean,
      default: false,
    },
    /**
     * Sizing for this input.
     * `small`
     */
    size: {
      type: String,
      default: null,
      validator: isInList(['small']),
    },
    /**
     * A note to be displayed below this input.
     */
    note: {
      type: String,
      default: null,
    },
  },

  computed: {
    listeners() {
      return {
        ...this.$listeners,
        // TODO: if we change 'change' to 'input' v-model
        // will work out of the box. Requires a refactor of
        // anything listening for the 'change' even however.
        change: (e) => this.$emit('change', e.target.value),
        input: (e) => this.$emit('change', e.target.value),
        focus: (e) => this.$emit('focus', e.target.value),
        blur: (e) => this.$emit('blur', e.target.value),
      };
    },
    computedClasses() {
      return {
        [`vue--input--${this.size}`]: this.size !== null,
      };
    },
    computedInputClasses() {
      return {
        [this.state]: this.state !== null,
        'vue--input__field--error': this.error,
      };
    },
    noteVariant() {
      return this.error ? 'error' : 'default';
    },
  },
};
</script>

<style lang="scss" scoped>
@import '~/assets/scss/utils';

.vue--input {
  $self: &;

  &--small {
    #{$self}__field {
      line-height: rem(22px);
    }

    #{$self}__label {
      line-height: rem(22px);
    }
  }

  &__field {
    @include fieldStyles();
    @include hoverState();

    &[disabled],
    &:disabled {
      @include disabledState();
    }

    &--error {
      @include errorState();
    }
  }

  &__note {
    margin-top: space(xs);
  }
}
</style>
