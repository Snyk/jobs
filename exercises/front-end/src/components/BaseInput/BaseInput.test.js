import { shallowMount } from '@vue/test-utils';
import BaseInput from '~/components/BaseInput/BaseInput';
import BaseInputNote from '~/components/BaseInputNote/BaseInputNote';

describe('BaseInput', () => {
  const wrapper = shallowMount(BaseInput);

  it('renders as expected', async () => {
    await wrapper.setProps({
      id: 'some-id',
      type: 'password',
      value: '**********',
      disabled: true,
      name: 'password',
      placeholder: 'Enter a password',
      label: 'Password',
      inline: true,
    });
    expect(wrapper).toMatchSnapshot();
  });

  describe('Computed Properties', () => {
    const wrapper = shallowMount(BaseInput, {
      propsData: {
        size: 'small',
      },
    });

    it('computes classes', () => {
      expect(wrapper.vm.computedClasses).toEqual({
        'vue--input--small': true,
      });
    });

    it('computes input classes', async () => {
      await wrapper.setProps({
        state: 'hover',
        error: true,
      });

      expect(wrapper.vm.computedInputClasses).toEqual({
        hover: true,
        'vue--input__field--error': true,
      });
    });
  });

  describe('With note', () => {
    const wrapper = shallowMount(BaseInput, {
      propsData: {
        note: 'Note',
      },
    });

    it('renders as expected', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('contains BaseInputNote', () => {
      expect(wrapper.findComponent(BaseInputNote).isVisible()).toBe(true);
      expect(wrapper.vm.noteVariant).toEqual('default');
    });

    it('displays error variant', async () => {
      await wrapper.setProps({
        error: true,
      });
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.vm.noteVariant).toEqual('error');
    });

    it('does not contain BaseInputNote when note prop not set', async () => {
      await wrapper.setProps({
        note: undefined,
      });
      expect(wrapper.findComponent(BaseInputNote).exists()).toBe(false);
    });
  });

  describe('Events', () => {
    const wrapper = shallowMount(BaseInput);

    it('emits a focus event', async () => {
      const input = wrapper.find('input');
      await input.trigger('focus');
      expect(wrapper.emitted('focus')).toBeTruthy();
    });

    it('emits a changed event', async () => {
      const input = wrapper.find('input');
      input.element.value = 'some text';
      await input.trigger('input');
      expect(wrapper.emitted('change')).toBeTruthy();
      expect(wrapper.emitted('change')[0]).toEqual(['some text']);
    });
  });
});
