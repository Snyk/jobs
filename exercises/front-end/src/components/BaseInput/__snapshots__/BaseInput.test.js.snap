// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseInput With note displays error variant 1`] = `
<layoutformelement-stub fillwidth="true" class="vue--input"> <input type="text" class="vue--input__field vue--input__field--error">
  <baseinputnote-stub variant="error" class="vue--input__note">
    Note
  </baseinputnote-stub>
</layoutformelement-stub>
`;

exports[`BaseInput With note renders as expected 1`] = `
<layoutformelement-stub fillwidth="true" class="vue--input"> <input type="text" class="vue--input__field">
  <baseinputnote-stub variant="default" class="vue--input__note">
    Note
  </baseinputnote-stub>
</layoutformelement-stub>
`;

exports[`BaseInput renders as expected 1`] = `
<layoutformelement-stub fillwidth="true" class="vue--input" inline="true"> <input type="password" class="vue--input__field" id="some-id" disabled="disabled" name="password" placeholder="Enter a password">
  <!---->
</layoutformelement-stub>
`;
